docs/rfc-1-staking-contract.adoc

```diff
@@ -0,0 +1,381 @@
+controlled by the governance.
+
+==== Stake top-up
+
+Top-up increases the amount of tokens locked in a delegation. This increases the
+probability of being chosen for work in the network but is only effective for
+future checks of the stake amount.
+
+Top-ups can be executed for native T stakes and for legacy KEEP/NU stakes.
+
+Native T stakers can only top-up their stakes with a liquid T.
+
+Existing KEEP and NU stakers wanting to execute a top-up have two options. One
+option is to wrap their KEEP/NU to T and then, execute a top-up in T staking
+contract. The second option is to execute a top-up in their legacy staking
+contracts and notify T staking contract about the fact their legacy stake
+increased.
+
+Effectively, it means that existing KEEP stakers can mix their legacy KEEP
+stakes with liquid T stakes. Similarly, existing NU stakers can mix their legacy
+NU stakes with liquid T stakes. This functionality adds some complexity to the
+staking contract but it puts existing KEEP/NU stakers in the same spot as new T
+stakers. Without it, existing stakers would not be able to top-up their stakes
+with T earned from operating in the network, so they would be in a worse spot
+than new T stakers allowed to top-up their stakes using earned T and this way
+increasing their rewards.
+
+Anyone can execute a stake top-up for an operator using a liquid T. Stake top-up
+does not automatically increase authorization levels for applications.
+
+==== Undelegating a stake
+
+The owner or operator may choose to undelegate the stake by submitting an
+undeleagation order. Stake delegation should be marked in the staking contract
+as undelegating. Always the entire stake is getting undelegated and there is a
+fixed, non-governable, two months undelegation period. All applications should
+be designed in such a way that a two-month undelegation period is sufficient to
+retire all groups undelegating operators are involved in. The operator is not
+eligible for any new work during the stake undelegation period but it is still
+required to finish all the pending work. Authorized applications are eligible
+to slash the operator for the entire undelegation period.
+
+==== Recovering a stake
+
+Once the undelegation period finishes, anyone - even a third party - can recover
+the stake. Staking contract sets the stake as recovered and returns all
+delegated tokens to the owner. Applications shouldn't be able to slash after
+undelegation is complete, even before the stake is returned. Information about
+owner, operator, and beneficiary, is retained in the contract in case some
+application rewards are still available for withdrawal.
+
+==== Keeping information in sync
+
+To avoid expensive calls to legacy staking contracts, it is assumed that
+information in T staking contract about the staked amount is always up-to-date.
+By design, this is always the case for native T stakers but might not be the
+case for legacy KEEP/NU stakers.
```

Comment:
I think no reason to cache whole stake amount, only allocation (for legacy contracts). Because we still don't want to slash more than allocated

Reply:
The KEEP staking contract does not support exclusive allocations, so any stake allocations on the T side may get slashed on the legacy KEEP contract. Because of this, the allocations have to be treated as "risk ceilings" where the real value that counts is `min(allocated, staked)`.

For example, Alice could start with 1,000,000 tokens and allocate them all to app X, then get slashed on KEEP and only have 900,000 tokens remaining while the allocation is unchanged at 1,000,000. As a result, we need the T staking contract to correctly calculate Alice's stake as 900,000 after being informed of the slashing.

If we need allocations to be exclusive (if Alice allocates 300,000 tokens to app Y she can only allocate 700,000 tokens to app Z) it needs to be treated as a special case that doesn't support legacy KEEP stakes.

Reply:
Allocations are not exclusive (#13) and they are already stored, I'm asking about stake amount in KEEP/NU. Even this amount is stored - it can't prevent from described situation and we will have sync/slash mechanism to prevent this.

Reply:
With non-exclusive allocations, it is still possible to run into trouble if we don't cache the stake amount locally.

If Alice has allocated the full 1M tokens to two different apps and gets slashed on one, the allocation on the other cannot be reduced to match because the gas cost of doing so would be unbounded with an increasing number of apps (due to having to iterate trough all possible apps and update their allocation if it's higher than the staked amount).

As a result, we need to check both the amount allocated for the app and the amount the operator actually has staked, and use the lower of the two values.

Reply:
Right now we consider 3 apps, iterate through dozens of apps still will be not so heavy. If remove iterations then app can't just wait calls with allocation changes so each time need to check contract again which leads to idea with callback won't work. Also this part is only for legacy stuff so I hope when we will have too much apps - we won't need in this optimizations

Reply:
Iterating through each app is not just an issue with legacy staking, but also any new ones that exclusively support T:

Updating an operator's weight in a sortition pool might cost 50,000 gas as a part of a larger transaction. If we iterate through e.g. 10 allocations and have to perform a callback to update the sortition pools for each of them, it would be 500,000 gas from the pool operations and around 60,000 gas from the allocations on the T staking contract. With 10 operators the cost would be 5.6M gas, and doing it to 100 operators would go way past the gas limit with 56M.

On the other hand, if we only update the stake and let the other applications go out of sync, we can handle 100 operators with around 600,000 gas.

If we need to slash a large number of operators at once, updating the applications becomes prohibitively expensive which is why the new sortition pool design contains a mechanism for incentivising syncing the pool and the staking contract without relying on the staking contract performing callbacks.

Reply:
Allowing stake information to go out of sync wrt to applications can break the rewards distribution model that we already agreed on. This is also one of the reasons we opened issue #14, precisely to disallow this possibility.. 

@eth-r Can you clarify on what scenarios there's a need to update 10 or 100 operators' weight at once? Also, I don't understand why sortition pools are being mentioned because they're application-specific (AFAIK, for tBTCv2), and as such, I don't think such restrictions apply to the general T staking model. Similarly, we don't need to consider a model with an unbounded number of apps per stake (we currently only have 3); quite the opposite, we probably should limit it to a sane number that allows to iterate when needed.

Reply:
If we consider a scenario where a tBTC or random beacon group has produced an unauthorised signature, we want to slash the entire group (100 members) because a majority must have been dishonest. Now all these 100 members have their stake reduced. With non-exclusive allocations, this potentially affects the allocation for every other application for every single member.

Sortition pools aren't tBTC-specific, we're also planning on using them for random beacon v2 and quite likely for any future apps with similar needs, because an improved version I'm currently working on solves practically every group selection problem we've had (apart from the one where bulk updating 100 operators costs 5 million instead of 500k gas).

If PRE needs to stay up to date in all circumstances we can afford to special-case it, but I would not tie us down to a model where everything has to stay in sync at all times especially when the number of apps where iteration with callbacks runs OOG can be as low as 2 (e.g. tBTCv2 + random beacon v2: 100 operators get slashed on one app and banned from its sortition pool (>5M gas), have their stake adjusted (>600k gas), which triggers updates on the other application's sortition pool (another >5M gas)).

Reply:
Authorization is just a risk ceiling and the eligible stake for the application is `min(authorization, stake)`. Specifically, the application should not treat the "allocation" (authorization) as a stake; otherwise, we are running into problems described here. 

Caching the entire stake amount feels simpler implementation-wise. For each new authorization (allocation), we wouldn't need to consult the legacy contract and just check the locally cached value assuming it is up to date. If it is not, someone will notify about this fact and the operator will get slashed.

There is an interesting corner case when authorization happens before someone has a chance to notify about discrepancy but I think we can fix it by allowing to notify per application and slashing multiple times. (?)

Reply:
Adding one call to authorization is not so expensive considering that authorization calls will be rare comparing other functions. Like for average longtime user it will be few times. And this fully remove issue with not syncing at authorization time. Caching this value and not syncing on slashing is bigger question, I agree